Generator/Demo Preset – Feedback for Vincent

olibrook opened the pull request.

Commit:
Update generator with a flag for --preset and make it self-documenting.

Commit:
Fix script name

rsexton404 committed:
Tweak help text to match Salesforce style

rsexton404 committed:
Revert unwanted Prettier formatting

rsexton404 reviewed (approved):
I like being able to pass in the preset as an option on the command line because it simplifies our Quick Start command (no environment variable required), and splitting the presets into public and private is nice. Although I prefer the way the prompts worked before, it's not very important to me. I feel like it offered a slightly better user experience because you did not have to consult the help text to know about the demo settings. It was also how John was expecting it to work, BTW.

Once this PR is finalized, can we release it on NPM as soon as possible so that I can update our Getting Started instructions? I have a large update to the docs that will be held up until this generator update is available.

rsexton404 commented on packages/pwa-kit-create-app/package.json:
```diff
@@ -24,7 +24,7 @@
   },
   "dependencies": {
-    "commander": "^2.20.0",
+    "commander": "^8.3.0",
```
This change appears to be causing tests to fail with the following error:

```
Package "pwa-kit-create-app" requires "commander@^8.3.0" but we've standardized on "commander@^2.20.0" in the monorepo. To fix this, change the version in pwa-kit-create-app to match "commander@^2.20.0."
```

vcua-mobify commented:
I am not super attached to the approach we take so this way of using `--preset demo-project` works for me.

I guess if I have any points to raise it would be the following:
* I am under the impression that the `--preset ` parameter is purposely supposed to be undocumented and is meant to be used only for testing (and it's says this in the top level comment in `create-mobify-app.js`). I don't have any issues with making `--preset` public but we should state that we are doing so and update the comment.
* The [proposed user workflow ](https://salesforce.quip.com/rgAgAomBHzbJ) for creating a project with demo settings describes using a prompt. Since this will be deviating from that flow, we might want to have @jennmills have a look at this as well before we release this to the public

rsexton404 commented:
@vcua-mobify The `preset` option is documented when you invoke the `--help` option. You can test it locally with this command:

```
node scripts/create-mobify-app-dev.js --help
```

rsexton404 committed:
Update comments to describe new environment variable

rsexton404 commented:
Vincent pointed out that the comments described the old `PRESET` environment variable as "purposely undocumented," so I updated the comments to reflect the current state of the tool. @olibrook Can you double check that the comments are accurate now?

olibrook commented:
@rsexton404 @vcua-mobify I think people are quickly convinced if I compare our stuff to other platforms. Here's what next does:

```
➜  pwa-kit git:(generator-with-demo-settings--feedback) npx create-next-app@latest --help
npx: installed 1 in 1.827s
Usage: create-next-app <project-directory> [options]

Options:
  -V, --version                      output the version number
  --ts, --typescript

    Initialize as a TypeScript project.

  --use-npm

    Explicitly tell the CLI to bootstrap the app using npm

  -e, --example [name]|[github-url]

    An example to bootstrap the app with. You can use an example name
    from the official Next.js repo or a GitHub URL. The URL can use
    any branch and/or subdirectory
```

Their [example projects live here]. So, our `--preset <name>` is the same as their `--example <name>`.

@vcua-mobify With respect to private/public questions, the env var was purposefully undocumented (by me!) because we didn't want to publicly support more than one preset. But with your change we do. And we keep talking about the need for more of them (Typescript, OneCommerce, Supermarkets, etc.). So, your new use-case is justification enough for making it public _now_.

@rsexton404 Excellent catch about the docs. I'll sense check those now, myself.

[example projects live here]: https://github.com/vercel/next.js/tree/canary/examples

Commit:
Review and update docs & preset names

olibrook commented:
@vcua-mobify @rsexton404 I reviewed and update the docs and the preset names. Here's what the CLI --help looks like now:

```
Usage: pwa-kit-create-app [options]

Generate a new PWA Kit project, optionally using a preset.

Examples:

  pwa-kit-create-app --preset "retail-react-app"
    Generate a project using custom settings by answering questions about a
    B2C Commerce instance.

    Use this preset to connect to an existing instance, such as a sandbox.

  pwa-kit-create-app --preset "retail-react-app-demo"
    Generate a project using the settings for a special B2C Commerce
    instance that is used for demo purposes. No questions are asked.

    Use this preset to try out PWA Kit.


Options:
  --outputDir <path>  Path to the output directory for the new project (default: "/Users/o.brook/repos/pwa-kit/packages/pwa-kit-create-app/pwa-kit-starter-project")
  --preset <name>     The name of a project preset to use (choices: "retail-react-app", "retail-react-app-demo", default: "retail-react-app", env: GENERATOR_PRESET)
  -h, --help          display help for command
```

Changes:

1. Renamed the preset `prompt` -> `retail-react-app`
2. Renamed the preset `demo` -> `retail-react-app-demo`.

The reasoning for that is that we _know_ we are immediately going to have to create more project templates. And we'll need "demo" versions of _those_ templates, too. So I want to make it obvious how to add eg. a preset for `1-commerce`, like this: `pwa-kit-create-app --preset "1-commerce"` or `pwa-kit-create-app --preset "1-commerce-demo"`.

My suggestion would be that anything that ends in `-demo` would be pre-configured to talk to a special pre-built backend.

olibrook commented:
@rsexton404 @vcua-mobify Reminder that we don't get opportunities to rename these presets, so we need to plan for growth and make sure we're happy enough with the names!

vcua-mobify commented:
Just leaving a comment here that @olibrook and I have synced up on this PR and on the remaining change to fix the CI tests. I'll be following up on this moving onward.

rsexton404 commented:
Love the new template names and naming scheme with the `-demo` suffix. Since we tend to use the term "template" a lot, should we rename `--preset` (and other associated variable names and strings) to `--template`? I know that @sejal-salesforce for one would like to promote the term "template" to describe the Retail React App (and others apps like it).

olibrook commented:
@rsexton404 I thought about `--template`. The only reason I stuck with `--preset` is that there are more than one, uh, "preset" for the same template. For me "template" corresponds to a folder – ie. the pwa folder. And the `retail-react-app` and `retail-react-app-demo` presets use the same template to generate a project.

vcua-mobify committed:
Revert commander version

vcua-mobify committed:
Update package-lock.json

vcua-mobify committed:
FIx lint errors

vcua-mobify commented:
@olibrook I've downgraded the commander version. I couldn't find a substitute for `choices` in the older version of commander so for now I've just typed out the available options as part of the --preset description.

Also, in the current iteration of the PR, in the edge scenario where the GENERATOR_PRESET env var and --preset arg are both present, the env var setting wins out. We can swap this around if we want to prioritize what is passed in --preset instead.

vcua-mobify commented:
Merging this in to #322

sejal-salesforce commented:
I see we closed and merged it, so I am late to the party. I still got questions though. @vcua-mobify what's the best way to invoke the generator from monorepo? 

@olibrook  preset is fine as it works for all types of errr.. presets - examples, templates, demos, starters, what not. So all is good there. And like the conventions around `-demo` suffix. I can see us doing the same with examples with `-example` suffix (and I'm getting ideas now 😄 ). What's the behaviour if there is no `--preset`? 

I do think we should drop `react` from the `retail-react-app` though.
